src/Agoda.Analyzers/AgodaCustom/AG0043NoBuildServiceProvider.cs

```diff
@@ -0,0 +1,101 @@
+using System.Collections.Immutable;
+using System.Linq;
+using Agoda.Analyzers.Helpers;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Agoda.Analyzers.AgodaCustom
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public class AG0043NoBuildServiceProvider : DiagnosticAnalyzer
+    {
+        public const string DIAGNOSTIC_ID = "AG0043";
+
+        private static readonly LocalizableString Title = new LocalizableResourceString(
+            nameof(CustomRulesResources.AG0043Title),
+            CustomRulesResources.ResourceManager,
+            typeof(CustomRulesResources));
+
+        private static readonly LocalizableString MessageFormat = new LocalizableResourceString(
+            nameof(CustomRulesResources.AG0043Title),
+            CustomRulesResources.ResourceManager,
+            typeof(CustomRulesResources));
+
+        private static readonly LocalizableString Description
+            = DescriptionContentLoader.GetAnalyzerDescription(nameof(AG0043NoBuildServiceProvider));
+
+        private static readonly DiagnosticDescriptor Descriptor = new DiagnosticDescriptor(
+            DIAGNOSTIC_ID,
+            Title,
+            MessageFormat,
+            AnalyzerCategory.CustomQualityRules,
+            DiagnosticSeverity.Error,
+            AnalyzerConstants.EnabledByDefault,
+            Description,
+            "https://github.com/agoda-com/AgodaAnalyzers/issues/190",
```

Comment:
This is a tricky one, should we add to docs some examples for typical use-cases and how to solve them without `BuildServiceProvider`? Where I've seen this used most of the time is during constructing services that have their own dependencies, like in the registration of a lot of the thick-clients for agoda services. I assume the right way would be to just pass an IServiceProvider along, rather than building it?